highlighters/main/main-highlighter.zsh

```diff
@@ -795,18 +789,72 @@ _zsh_highlight_main_highlighter_check_path()
   return 1
 }
 
+# Highlight an argument and possibly special chars in quotes
+_zsh_highlight_main_highlighter_highlight_argument()
+{
+  local i
+
+  _zsh_highlight_main_add_region_highlight $start_pos $end_pos default
+  for (( i = 1 ; i <= end_pos - start_pos ; i += 1 )); do
+    case "$arg[$i]" in
+      "\\") (( i += 1 )); continue;;
+      "'") _zsh_highlight_main_highlighter_highlight_single_quote $i; (( i = REPLY ));;
+      '"') _zsh_highlight_main_highlighter_highlight_double_quote $i; (( i = REPLY ));;
+      '$')
+        if [[ $arg[i+1] == "'" ]]; then
+          _zsh_highlight_main_highlighter_highlight_dollar_quote $i
+          (( i = REPLY ))
+        elif [[ $arg[i+1] == [\^=~#+] ]]; then
+          while [[ $arg[i+1] == [\^=~#+] ]]; do
+            (( i += 1 ))
+          done
+          if [[ $arg[i+1] == [*@#?-$!] ]]; then
+            (( i += 1 ))
+          fi
+        elif [[ $arg[i+1] == [*@#?-$!] ]]; then
+          (( i += 1 ))
+        fi;;
+      [*?])
+        if $highlight_glob; then
+          _zsh_highlight_main_add_region_highlight $start_pos $end_pos globbing
+          break
+        fi;;
+      *) continue;;
+    esac
+  done
+}
+
+# Highlight single-quoted strings
+_zsh_highlight_main_highlighter_highlight_single_quote()
+{
+  local arg1=$1 i q=\'
+  local -a highlights
+  i=$arg[(ib:arg1+1:)$q]
+
+  if [[ $useroptions[rcquotes] == on ]]; then
+    while [[ $arg[i+1] == "'" ]]; do
+      highlights+=($(( start_pos + i - 1 )) $(( start_pos + ++i )) rc-quote)
```

Comment:
Parentheses around `++i`?  Better yet, break incrementing `i` out to a separate line, since it's not _obviously_ correct to have `++i` on the same line as `i - 1`?